src/mlpack/core/data/dict_encoding_impl.hpp

```diff
@@ -0,0 +1,72 @@
+/**
+ * @file dict_encoding_impl.hpp
+ * @author Jeffin Sam
+ *
+ * Implementation of dictionary encoding functions.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_DATA_DICT_ENCODING_IMPL_HPP
+#define MLPACK_CORE_DATA_DICT_ENCODING_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "dict_encoding.hpp"
+#include <string>
+#include <vector>
+#include <unordered_map>
+
+namespace mlpack {
+namespace data {
+
+/**
+ * Dictionary Encoding
+ * here we simply assign a word (or a character) to a numeric index
+ * and treat the dataset as categorical.
+ *
+ * @param vector of documents.
+ * @param mapping of string to their encoded number.
+ * @param output matrix.
```

Comment:
Each `@param` macro shuold have the name of the parameter first, then a full sentence of documentation afterwards.  So, e.g., `@param strings Input vector of documents.`  Although---isn't this just an input vector of strings, not documents?  I'm unclear on whether the goal is to convert a list of strings to dictionary encoding or to convert a list of lists of characters to dictionary encoding.  The code looks like every `std::string` should contain multiple words?

It might be nice if we had some ability to change that functionality---some users might want to encode a string like `GATTACA` as seven different values (for instance, from biology); others might want words that are separated by spaces.  Others still might want something that tokenizes differently.